### ISSUE TITLE
Infinite recursion error with moto 5.0.8

### ISSUE THREAD

**Issue by sodul** [getmoto/moto#7730]
After upgrading our CI container from moto 5.0.7 to 5.0.8 we started seeing an infinite recursion error in one of our test cases.

```
Ran 1 test in 2.552s

FAILED (errors=1)
/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/botocore/auth.py:419: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  datetime_now = datetime.datetime.utcnow()

Error
Traceback (most recent call last):
  File "/home/aws/test/test_vpc.py", line 240, in test_create_vpc_flow_logs_in_account_moto
    vpc_func.create_vpc_flow_logs_in_account(self.ctx_obj, dst_bucket='test-bucket')
  File "/home/aws/vpc_func.py", line 108, in create_vpc_flow_logs_in_account
    completed_future.result()
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/aws/vpc_func.py", line 40, in create_flow_log
    response = client.create_flow_logs(
               ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/botocore/client.py", line 565, in _api_call
    return self._make_api_call(operation_name, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/botocore/client.py", line 1001, in _make_api_call
    http, parsed_response = self._make_request(
                            ^^^^^^^^^^^^^^^^^^^
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/botocore/client.py", line 1027, in _make_request
    return self._endpoint.make_request(operation_model, request_dict)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/botocore/endpoint.py", line 119, in make_request
    return self._send_request(request_dict, operation_model)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/botocore/endpoint.py", line 202, in _send_request
    while self._needs_retry(
          ^^^^^^^^^^^^^^^^^^
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/botocore/endpoint.py", line 354, in _needs_retry
    responses = self._event_emitter.emit(
                ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/botocore/hooks.py", line 412, in emit
    return self._emitter.emit(aliased_event_name, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/botocore/hooks.py", line 256, in emit
    return self._emit(event_name, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/botocore/hooks.py", line 239, in _emit
    response = handler(**kwargs)
               ^^^^^^^^^^^^^^^^^
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/botocore/retryhandler.py", line 207, in __call__
    if self._checker(**checker_kwargs):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/botocore/retryhandler.py", line 284, in __call__
    should_retry = self._should_retry(
                   ^^^^^^^^^^^^^^^^^^^
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/botocore/retryhandler.py", line 307, in _should_retry
    return self._checker(
           ^^^^^^^^^^^^^^
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/botocore/retryhandler.py", line 363, in __call__
    checker_response = checker(
                       ^^^^^^^^
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/botocore/retryhandler.py", line 247, in __call__
    return self._check_caught_exception(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/botocore/retryhandler.py", line 416, in _check_caught_exception
    raise caught_exception
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/botocore/endpoint.py", line 278, in _do_get_response
    responses = self._event_emitter.emit(event_name, request=request)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/botocore/hooks.py", line 412, in emit
    return self._emitter.emit(aliased_event_name, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/botocore/hooks.py", line 256, in emit
    return self._emit(event_name, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/botocore/hooks.py", line 239, in _emit
    response = handler(**kwargs)
               ^^^^^^^^^^^^^^^^^
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/moto/core/botocore_stubber.py", line 37, in __call__
    response = self.process_request(request)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/moto/core/botocore_stubber.py", line 84, in process_request
    status, headers, body = method_to_execute(
                            ^^^^^^^^^^^^^^^^^^
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/moto/core/responses.py", line 290, in dispatch
    return cls()._dispatch(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/moto/core/responses.py", line 501, in _dispatch
    return self.call_action()
           ^^^^^^^^^^^^^^^^^^
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/moto/core/responses.py", line 587, in call_action
    response = method()
               ^^^^^^^^
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/moto/ec2/responses/flow_logs.py", line 23, in create_flow_logs
    flow_logs, errors = self.ec2_backend.create_flow_logs(
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/moto/ec2/models/flow_logs.py", line 231, in create_flow_logs
    s3_backends[self.account_id][self.partition].get_bucket(arn)  # type: ignore[attr-defined]
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/moto/s3/models.py", line 1848, in get_bucket
    return s3_backends[account_id][self.partition].get_bucket(bucket_name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/moto/s3/models.py", line 1848, in get_bucket
    return s3_backends[account_id][self.partition].get_bucket(bucket_name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/moto/s3/models.py", line 1848, in get_bucket
    return s3_backends[account_id][self.partition].get_bucket(bucket_name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  [Previous line repeated 970 more times]
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/moto/s3/models.py", line 1845, in get_bucket
    if not s3_allow_crossdomain_access():
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/.pyenv/versions/3.12.3/lib/python3.12/site-packages/moto/settings.py", line 76, in s3_allow_crossdomain_access
    return os.environ.get("MOTO_S3_ALLOW_CROSSACCOUNT_ACCESS", "true").lower() == "true"
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen _collections_abc>", line 807, in get
  File "<frozen os>", line 682, in __getitem__
RecursionError: maximum recursion depth exceeded
```

**Comment by sodul** [comment]
I looked at the code under test, which is in a private repositories so I cannot share all the details publicly here.
Something different about that code is that it is called under:
```
 with concurrent.futures.ThreadPoolExecutor(max_workers=max_workers) as executor:
```
Then the infinite recursion happens when we call:
```
        for completed_future in concurrent.futures.as_completed(futures):
            completed_future.result()
```

The boto3 code call was create_flow_logs() for EC2.

**Comment by bblommers** [comment]
Hi @sodul, thanks for raising this. I've just pushed a fix for this - are you able to reproduce it with `moto >= 5.0.9.dev8`?

**Comment by sodul** [comment]
Tested:
- 5.0.7: works
- 5.0.8: fails with the infinite recursion
- 5.0.9.dev8: works

Thanks for the quick fix!

**Comment by bblommers** [comment]
Great, thanks for confirming that @sodul.

I've just released 5.0.9, so I'll close this then.
